metricbeat/helper/module.go

```diff
@@ -1,130 +1,144 @@
 	}
 }
 
-// Registers moudle with central registry
-func (m *Module) Register() {
-	Registry.AddModule(m)
-}
+// Starts the given module
+func (m *Module) Start(b *beat.Beat) {
 
-// Add metric to module
-func (m *Module) AddMetric(metricSet *MetricSet) {
-	m.MetricSets[metricSet.Name] = metricSet
-}
+	defer func() {
+		if r := recover(); r != nil {
+			logp.Err("Module %s paniced and stopped running. Reason: %+v", m.name, r)
+		}
+	}()
 
-// Interface for each module
-type Moduler interface {
-	Setup() error
-}
+	if !m.config.Enabled {
+		logp.Debug("helper", "Not starting module %s as not enabled.", m.name)
+		return
+	}
 
-// Base configuration for list of modules
-type ModulesConfig struct {
-	Modules map[string]ModuleConfig
-}
+	logp.Info("Setup moduler: %s", m.name)
+	err := m.moduler.Setup()
+	if err != nil {
+		logp.Err("Error setting up module: %s. Not starting metricsets for this module.", err)
+		return
+	}
 
-// Base module configuration
-type ModuleConfig struct {
-	Hosts      []string
-	Period     string
-	MetricSets map[string]MetricSetConfig `yaml:"metricsets"`
-}
+	// Setup - Create metricSets for the module
+	for _, metricsetName := range m.config.MetricSets {
+		metricSeter := Registry.MetricSeters[m.name][metricsetName]
+		// Setup
+		err := metricSeter.Setup()
+		if err != nil {
+			logp.Err("Error happening during metricseter setup: %s", err)
+		}
 
-// Helper functions to easily access default configurations
-func (m *Module) GetPeriod() (time.Duration, error) {
-	return time.ParseDuration(m.BaseConfig.Period)
-}
+		metricSet := NewMetricSet(metricsetName, metricSeter, m.config)
+		m.metricSets[metricsetName] = metricSet
+	}
+
+	// Setup period
+	period, err := time.ParseDuration(m.config.Period)
+	if err != nil {
+		logp.Info("Error in parsing period of metric %s: %v", m.name, err)
+	}
 
-func (m *Module) GetHosts() []string {
-	return m.BaseConfig.Hosts
+	// If no period set, set default
+	if period == 0 {
+		logp.Info("Setting default period for metric %s as not set.", m.name)
+		period = 1 * time.Second
+	}
+
+	// TODO: Improve logging information with list (names of metricSets)
+	logp.Info("Start Module %s with metricsets %v and period %v", m.name, m.metricSets, period)
+
+	go m.Run(period, b)
 }
 
-// Loads the configurations specific config.
-// This needs the configuration object defined inside the module
-func (m *Module) LoadConfig(config interface{}) {
-	bytes, err := yaml.Marshal(m.RawConfig)
+func (m *Module) Run(period time.Duration, b *beat.Beat) {
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
 
-	if err != nil {
-		logp.Err("Load module config error: %v", err)
+	for {
+		select {
+		case <-m.done:
+			logp.Info("Stopping module %s with metricsets ... TODO", m.name)
+			return
+		case <-ticker.C:
+		}
+
+		// Execute
+		for _, set := range m.metricSets {
+			// TODO: Aggregate events per module and send as bulk?
+			// Move execution part to module?
+			go m.FetchMetricSets(b, set)
```

Comment:
That would imply that if one of the metricsets is hanging, it would also affect the others? I initially thought about using the module to also aggregate the events and send them as one event but decided agains it because I don't think that one metricset should be able to block an other one.